hoodie-spark/src/main/scala/com/uber/hoodie/HoodieStreamingSink.scala

```diff
@@ -0,0 +1,57 @@
+/*
+ *  Copyright (c) 2017 Uber Technologies, Inc. (hoodie-dev-group@uber.com)
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *           http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *
+ */
+package com.uber.hoodie
+
+import org.apache.spark.sql.{DataFrame, SQLContext, SaveMode}
+import org.apache.spark.sql.execution.streaming.Sink
+import org.apache.spark.sql.streaming.OutputMode
+import org.apache.log4j.LogManager
+
+class HoodieStreamingSink(sqlContext: SQLContext,
+                          options: Map[String, String],
+                          partitionColumns: Seq[String],
+                          outputMode: OutputMode)
+  extends Sink
+    with Serializable {
+  @volatile private var latestBatchId = -1L
+
+  private val log = LogManager.getLogger(classOf[HoodieStreamingSink])
+
+  private val mode =
+    if (outputMode == OutputMode.Append()) {
+      SaveMode.Append
+    } else {
+      SaveMode.Overwrite
+    }
+
+  override def addBatch(batchId: Long, data: DataFrame): Unit = {
+    if (batchId <= latestBatchId) {
+      log.info(s"Skipping already committed batch $batchId")
+    } else {
+      HoodieSparkSQLWriter.write(
+        sqlContext,
+        mode,
+        options,
+        data)
+      latestBatchId = batchId
```

Comment:
would we update the latestBatchId even if the write failed? Make `write` return boolean or throw an exception? I think Spark will retry a failing batch again  and I am concerned we will skip that batch of input records?

Reply:
Another thought, should we make `batchId` = `commitTime` so its very easy to reason about? i.e pass it on and make the HoodieWriteClient etc take that instead of using the default YYYYMMDDHHMMSS format.. As long as batchId is increasing, we should be fine.. wdyt?

Reply:
I copied this from KafkaSink - my understanding is that this is trying to avoid spark's retry on a committed micro-batch. batchId comparison should be good enough?

I like the idea to make sure the write is successful, currently only error messages in job log? (not sure how batch jobs management system in uber handle those errors) We might want to throw exceptions with certain retries to make sure data integrity.

A couple of concerns if using batchId as commitTime: 
1. The [spark generated batchId](https://github.com/apache/spark/blob/v2.1.0/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala#L94) is generated by spark and it start from 0 when streaming job starts and increment every each micro batch. To make sure this is working correctly no matter whether we have streaming job restarted or not, we might need to find out the latest saved commitTime from the existing table every time we restart streaming job and add it to the spark generated batchId to generate new commitTime.

2. The default format has more info about the commit (the utc timestamp) and internally we are using it to generate ingest_time as our offset for downstream jobs, so when we try to recalculate some old data, we can specify a certain time range of ingest_time. This will be harder if we use batchId as the offset.